### ISSUE TITLE
[RFC] Support DAPO training for Qwen3-Omni Thinker

### ISSUE THREAD

**Issue by SamitHuang** [verl-project/verl-omni#446]
### 1. Motivation

Qwen3-Omni Thinker is already a first-class omni policy in `verl-omni`. The README currently lists **DPO** and **GSPO** as the supported algorithms. [DAPO](https://arxiv.org/abs/2503.14476) (Decoupled Clip and Dynamic sAmpling Policy Optimization) is the production-scale GRPO-family recipe in upstream `verl` for long-CoT / math RL: clip-higher, token-level policy gradient, overlong reward shaping, and dynamic sampling.

There is no Thinker DAPO recipe today. The GSPO V1 launcher (`examples/gspo_trainer/qwen3_omni/run_qwen3_omni_thinker_gspo_lora_v1.sh`) is a sequence-level GSPO setup (`policy_loss.loss_mode=gspo`, `clip_ratio_{low,high}=3e-4/4e-4`, `loss_agg_mode=seq-mean-token-mean`, `reward.reward_manager.name=naive`). The legacy GSPO YAML sets `reward.reward_manager.name=dapo`, which only selects the **DAPO reward manager** (overlong buffer / math parsing). That is not the DAPO **algorithm**.

Supporting DAPO on Thinker is useful because:

1. Users already ask for it as the default long-horizon RL recipe on Qwen3-class models.
2. The Thinker adapter is algorithm-agnostic: DAPO should be a config + recipe + trainer-loop gap, not a new model adapter.
3. It gives a token-level GRPO baseline next to GSPO, which is needed to compare sequence-level vs token-level clipping on multimodal long responses.

This RFC proposes **Thinker-only DAPO** on the existing V1 omni path (`verl_omni.trainer.main_omni`, `OmniPPOTrainerSync`, vLLM-Omni AR rollout). It does not add a new trainer class.

### 2. Current baseline and gap

| Piece | Status |
| --- | --- |
| `OmniModelBase` / Qwen3-Omni Thinker training adapter | Available. Algorithm selection is `actor.policy_loss.loss_mode` + `algorithm.adv_estimator`. |
| V1 omni trainer (`omni_sync` → `PPOTrainerSync`) | Available. GSPO LoRA GPU and full-param NPU recipes exist. |
| vLLM-Omni AR rollout + LoRA weight sync | Available (`output_mode=ar`, `pipeline_name=qwen3_omni_moe`). |
| Upstream `verl` DAPO knobs | Available in `verl` (`clip_ratio_low/high`, `loss_agg_mode=token-mean`, `reward_manager=dapo`, `algorithm.filter_groups`). |
| Thinker DAPO recipe / docs / smoke | **Missing.** |
| `OmniAlgoConfig` fields for DAPO | **Incomplete.** The dataclass does not declare `filter_groups`, `use_kl_in_reward`, or `kl_ctrl`. |
| Dynamic sampling in the omni V1 loop | **Unverified.** Need to confirm `PPOTrainerSync` actually re-samples after group filtering. |

GSPO vs DAPO (do not mix):

| Knob | Current Thinker GSPO | DAPO (this RFC) |
| --- | --- | --- |
| `policy_loss.loss_mode` | `gspo` | `vanilla` (token-level PPO clip) |
| `clip_ratio_low / high` | `3e-4 / 4e-4` | `0.2 / 0.28` (clip-higher) |
| `loss_agg_mode` | `seq-mean-token-mean` | `token-mean` |
| `use_kl_loss` / `use_kl_in_reward` | false / false | false / false |
| `adv_estimator` | `grpo` | `grpo` |
| Reward manager | `naive` (V1) or `dapo` name-only (legacy) | `dapo` + overlong buffer |
| Group filtering | off | `algorithm.filter_groups.enable=true` (Phase 3) |

### 3. Goals and non-goals

**Goals**

1. First-class Thinker-only DAPO on the V1 omni trainer, reusing the existing adapter and vLLM-Omni AR rollout.
2. A documented GPU LoRA recipe (gsm8k) that is clearly **not** GSPO with a DAPO reward-manager name.
3. Enable the four DAPO components from the paper / `verl` recipe: clip-higher, token-level PG, overlong shaping, dynamic sampling.
4. CPU tests for config/contract, plus a tiny-random e2e smoke mirroring the GSPO V1 smoke.
5. README / docs row: Qwen3-Omni-Thinker × DAPO.

**Non-goals for the first release**

- Talker, code2wav, or full thinker+talker speech RL.
- Combining GSPO sequence clips with DAPO token-mean (incompatible).
- A new `OmniDAPOTrainer` class.
- Reproducing the DAPO paper AIME-2024 numbers on Qwen3-Omni.
- Megatron Thinker DAPO (can follow #100 once the FSDP V1 path is proven).

### 4. Proposed design

DAPO is a **policy-gradient configuration** of the existing Thinker stack.

```text
parquet → OmniModelBase (thinker) → vLLM-Omni AR rollout (n group samples)
        → DAPO reward manager (+ overlong) → GRPO advantages
        → vanilla clip-higher + token-mean PG → FSDP2 LoRA update → weight sync
```

#### 4.1 Config contract

Add `examples/dapo_trainer/qwen3_omni/` with a launcher that mirrors the GSPO V1 script except for algorithm knobs:

```bash
actor_rollout_ref.actor.policy_loss.loss_mode=vanilla
actor_rollout_ref.actor.clip_ratio_low=0.2
actor_rollout_ref.actor.clip_ratio_high=0.28
actor_rollout_ref.actor.clip_ratio_c=10.0
actor_rollout_ref.actor.loss_agg_mode=token-mean
actor_rollout_ref.actor.use_kl_loss=false
actor_rollout_ref.actor.entropy_coeff=0
algorithm.adv_estimator=grpo
algorithm.use_kl_in_reward=false
reward.reward_manager.source=register
reward.reward_manager.name=dapo
# overlong buffer (Hydra keys as used by current verl DAPO recipes)
# +reward.reward_kwargs.overlong_buffer_cfg.enable=true
# +reward.reward_kwargs.overlong_buffer_cfg.len=4096
# +reward.reward_kwargs.overlong_buffer_cfg.penalty_factor=1.0
```

Keep Thinker-only module filters unchanged:

```text
exclude_modules=".*talker.*|.*code2wav.*|.*code_predictor.*|.*visual.*|.*audio_tower.*"
actor.freeze_vision_tower=true
```

#### 4.2 `OmniAlgoConfig` pass-through

`omni_trainer.yaml` sets `algorithm._target_=verl_omni.trainer.config.OmniAlgoConfig`. That dataclass currently only has `trainer_type`, `sample_source`, `paired_preference`, `adv_estimator`, `norm_adv_by_std_in_grpo`, `global_std`. DAPO needs at least:

- `use_kl_in_reward`
- `kl_ctrl` (even if unused)
- `filter_groups` (`enable`, `metric`, `max_num_gen_batches`)

Either extend `OmniAlgoConfig` with these fields (preferred) or stop replacing the upstream `AlgoConfig` for online policy-gradient runs.

#### 4.3 Dynamic sampling

Reuse `verl`'s group filter: drop groups whose metric is uniform (all 0 or all 1), keep sampling with `data.gen_batch_size` until `data.train_batch_size` qualified prompts are collected.

This is **Phase 3**. Phase 1 ships DAPO **without** dynamic sampling (same split as `verl`'s `run_dapo_wo_ds_*.sh`), because the omni V1 loop may not yet re-enter generation after filtering. If `PPOTrainerSync` already implements it, Phase 3 is config-only.

#### 4.4 Reward

Use the registered `dapo` reward manager with overlong shaping. Phase 1 can keep the existing math accuracy scorer used by GSPO gsm8k. Multimodal rewards (MMK12 / AVQA) reuse the GSPO dataset + reward path; do not invent a new judge for the first DAPO milestone.

### 5. Implementation plan

- [ ] **Phase 0 — contract and inventory**
  - Document GSPO vs DAPO knobs; confirm which DAPO fields survive `OmniAlgoConfig` conversion.
  - Confirm whether `PPOTrainerSync` / `OmniPPOTrainerSync` implements `filter_groups` re-sampling.
  - Confirm V1 `reward.reward_manager.name=dapo` plus overlong kwargs actually apply on the omni reward loop (not only the legacy `main_ppo` path).
  - **Exit:** a short design note in this issue: which knobs are config-only vs trainer-code.

- [ ] **Phase 1 — DAPO without dynamic sampling (GPU LoRA, gsm8k)**
  - Add `examples/dapo_trainer/qwen3_omni/run_qwen3_omni_thinker_dapo_lora_v1.sh`.
  - `loss_mode=vanilla`, clip-higher, `token-mean`, no KL, GRPO advantages.
  - Tiny-random e2e smoke next to `tests/special_e2e/run_gspo_qwen3_omni_thinker_lora_v1_smoke.sh`.
  - CPU tests for the Hydra contract (wrong `loss_mode=gspo` + DAPO clips must fail or warn).
  - **Exit:** 2-step smoke on tiny-random Thinker; one short LoRA run on real `Qwen3-Omni-30B-A3B-Instruct` gsm8k with stable loss / grad-norm / reward.

- [ ] **Phase 2 — DAPO reward manager + overlong buffer**
  - Wire overlong shaping on the V1 reward path.
  - Log `overlong` / truncated-response metrics.
  - **Exit:** overlong penalty changes the reward on truncated dummy responses in smoke/CPU tests.

- [ ] **Phase 3 — dynamic sampling**
  - Enable `algorithm.filter_groups` through `OmniAlgoConfig` and the V1 generation loop.
  - If upstream `PPOTrainerSync` lacks the loop, add the smallest omni-side hook rather than a new trainer.
  - **Exit:** a batch of all-correct / all-wrong groups is dropped; trainer keeps generating until `train_batch_size` or `max_num_gen_batches`.

- [ ] **Phase 4 — multimodal recipes**
  - MMK12 and AVQA launchers, GPU LoRA first; NPU full-param only if the GSPO NPU recipe still applies with DAPO knobs.
  - **Exit:** one smoke or short run per modality recipe.

- [ ] **Phase 5 — docs and support matrix**
  - Example README, docs page, README algorithm table row `Qwen3-Omni-Thinker × DAPO`.
  - Explicit warning: `reward_manager.name=dapo` ≠ DAPO algorithm.

Phase 0 is the prerequisite. Phase 1 is the first user-visible milestone. Phase 3 is the completeness gate for calling the recipe “DAPO” rather than “GRPO + clip-higher”.

### 6. Validation plan

- CPU: Hydra DAPO vs GSPO contracts; `OmniAlgoConfig` keeps `filter_groups`; overlong penalty unit test.
- Tiny-random GPU/NPU smoke: 2 steps, LoRA, vLLM-Omni AR, DAPO knobs, no crash, finite loss.
- Short gsm8k LoRA run on 30B-A3B Thinker (same hardware class as GSPO V1: 4×H800). Compare against the existing GSPO recipe on reward, entropy, clip fraction, and response length — not as a quality bake-off, but to prove the token-level path is live.
- Dynamic-sampling unit/e2e: synthetic all-0 / mixed groups.
- Regression: GSPO V1 smoke still passes (no shared default mutation).

### 7. Major risks

1. **Config stripping:** `OmniAlgoConfig` may drop DAPO fields silently.
2. **Trainer loop:** V1 `omni_sync` may not implement dynamic sampling; shipping only clip-higher without calling that out would mislabel the recipe.
3. **Reward path:** V1 currently uses `naive`; `dapo` + overlong may only be wired on the deprecated `main_ppo` path.
4. **Clip semantics:** copying GSPO’s `3e-4/4e-4` into a vanilla token-level loss is incorrect and will under-update.
5. **Length / multimodal cost:** `token-mean` + long AVQA/MMK12 responses changes the loss scale vs GSPO; lr / batch may need a retune, not a 1:1 copy of GSPO hyperparameters.
6. **MoE train/infer mismatch:** same as GSPO; DAPO does not fix router replay.

### 8. Related work

- [DAPO paper](https://arxiv.org/abs/2503.14476) and [verl DAPO docs](https://verl.readthedocs.io/en/latest/algo/dapo.html)
- verl-omni [#232](https://github.com/verl-project/verl-omni/issues/232) — Formal Omni-Model Support (adapter is algorithm-agnostic)
- verl-omni [#183](https://github.com/verl-project/verl-omni/issues/183) / [#113](https://github.com/verl-project/verl-omni/pull/113) — Thinker GSPO + LoRA
- verl-omni [#100](https://github.com/verl-project/verl-omni/issues/100) — Thinker RL with Megatron (out of scope here)
- `examples/gspo_trainer/qwen3_omni/` — current Thinker PG baseline

Duplicate check: searched open issues/PRs for `DAPO` + `Qwen3-Omni`. Existing RFCs cover GSPO/FSDP/Megatron/omni V1, not a Thinker DAPO recipe.

AI assistance (Cursor Grok) was used to draft this RFC. A human submitter will own the discussion and any follow-up implementation.

**Comment by WenzheWang** [comment]
Hi @SamitHuang, I'd like to claim Phase 0 and Phase 1 of this RFC. I recently contributed #429 and #430 and would like to continue on the Qwen3-Omni trainer path.

I'll keep the first implementation independently reviewable:

- audit and report which DAPO fields survive `OmniAlgoConfig`, and whether the V1 reward and dynamic-sampling paths are actually wired;
- add the Thinker DAPO-without-dynamic-sampling GPU LoRA recipe;
- add CPU config/contract tests and the tiny-random two-step smoke;
- keep overlong reward integration, dynamic sampling, multimodal recipes, and NPU follow-ups out of PR1.

Before opening code, I'll post the Phase 0 inventory findings here. Please let me know if this slice conflicts with ongoing work or if you prefer extending `OmniAlgoConfig` versus retaining upstream `AlgoConfig` for online policy-gradient runs.

**Comment by OnePunchMonk** [comment]
Hi @SamitHuang @WenzheWang, I'd like to claim Phases 2–5 of this RFC (DAPO reward manager + overlong buffer, dynamic sampling, multimodal recipes, docs).

 I'll hold off starting until #456 (Phase 0+1) lands, since Phase 2 wires overlong shaping onto the V1 reward path Phase 1 establishes, and Phase 3 depends on the OmniAlgoConfig extensions from Phase 1. In the meantime, I'll track #456's review to base my branch on its head once merged (or its final revised commit if useful to start early on Phase 2's design).

 @WenzheWang please let me know if you're already planning to continue into Phase 2+ yourself once #456 merges; happy to defer if so.

**Comment by WenzheWang** [comment]
Thanks @OnePunchMonk — happy for you to take the remaining phases. I'll focus on getting #456 through review and avoid duplicating Phase 2/3 work.

A scope update from the review revisions:

- `OmniAlgoConfig` now inherits upstream `AlgoConfig`, with tests preserving enabled `filter_groups`, KL-control, and rollout-correction settings.
- The public recipe now uses GPU LoRA on AVQA with `naive` + `choice_reward`; dynamic sampling and overlong shaping remain disabled.
- The AVQA launcher and baseline docs are already included, so Phases 4/5 should build on those rather than recreate them.

The current head is `c4671984`, with the requested [single-seed AVQA validation curve](https://github.com/verl-project/verl-omni/pull/456#issuecomment-5449447190) posted; human re-review and maintainer-triggered CI are still pending.

Starting with overlong reward integration, then group-filtering/refill validation, sounds good. Please use the merged main as the implementation base; design discussion can proceed against the current PR meanwhile. We can keep #446 open to track the remaining work.
